cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceTaskContextProvider.java

```diff
@@ -207,10 +229,15 @@ public BasicMapReduceTaskContext load(ContextCacheKey key) throws Exception {
     private final Configuration configuration;
 
     private ContextCacheKey(TaskAttemptContext context) {
-      this.taskAttemptID = context.getTaskAttemptID();
-      this.configuration = context.getConfiguration();
+      this(context.getTaskAttemptID(), context.getConfiguration());
+    }
+
+    private ContextCacheKey(@Nullable TaskAttemptID taskAttemptID, Configuration configuration) {
```

Comment:
Am I understanding this correctly? 

For a partitioner or a comparator, the taskAttemptId will be null. That is, there is no metrics collection, right? 

Also, in standalone, the partitioner and comparator are instantiated by every mapper, is that right? If so, then that instance is cached and all mappers will share the same instance of the context. Can that lead to problems? For example, would it get initialized multiple times?